### DISCUSSION TITLE
getStaticProps redirect in production

### DISCUSSION THREAD

**Discussion by lmatheus** [vercel/next.js#21593] [Help]
I currently have some issues redirecting inside getStaticProps in production (**Vercel**)

**Context:**
I have a CMS where I change the path of a page. 

**Locally while developing:**
when hitting the old route the page is redirected correctly to the new route

**In production (**Vercel**)**
when hitting the old route, no redirection occurs and the page is still rendered (from the cache I believe), but also in the new route. This remains after hitting the old route a few times.

I have this setup on a `[...slug]` route in the root of the pages folder and no redirection set in `next.config.js`

```
export async function getStaticProps({ params }) {
  const path = params.path;
  // get page
  const page = await getPageByPath(path);

  // if there is no page check if there is a redirection set from the API
  if (!page) {
    const redirect = await getRedirect(path);
    //  redirect if exists with { source: xxxx, destination: xxxx, permanent: true/false }
    if (redirect) {
      return {
        redirect: {
          ...redirect,
        },
      };
    }
  }

  // if there is not page and no redirection force a notFound
  if (!page) {
    return {
      notFound: true,
    };
  }

  // otherwise continue as normal
  return {
    props: {
      page,
    },
    revalidate: 1,
  };
}

export async function getStaticPaths() {
  const data = await getAllPages();

  const paths = generateStaticPaths(data?.items);

  return {
    paths,
    fallback: 'blocking',
  };
}

export default Page;
```

Anything that I might be doing here that would change the behavior after built and in production?

**Comment by michrome** [comment]
You should configure the redirects via `next.config.js`. From https://nextjs.org/docs/basic-features/data-fetching#getstaticprops-static-generation -

> **Note**: Redirecting at build-time is currently not allowed and if the redirects are known at build-time they should be added in [`next.config.js`](https://nextjs.org/docs/api-reference/next.config.js/redirects).

**Reply by lmatheus** [reply]
Would this mean that the only way to make new redirects effective, I would have to rebuild the app manually when there is a route is changed for a page on a CMS and redirects come from an API?

**Reply by michrome** [reply]
Yes, a rebuild would be required. If the paths/pages are changing between builds, then they might not be static. If that's the case, you might be better off with `getServerSideProps`.

If they're changing _infrequently_, you could use a [Deploy Hook](https://vercel.com/docs/more/deploy-hooks#) to trigger a rebuild, as opposed to doing it manually.

**Reply by lmatheus** [reply]
Thank you for clarifying @michrome, this was quite helpful. I will add redirects in `next.config.js` and setup the deploy hook
